File: app/src/main/java/com/aaron/yespdf/preview/BookmarkAdapter.java
package com.aaron.yespdf.preview;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.aaron.yespdf.R;
import com.aaron.yespdf.common.widgets.ImageTextView;
import com.blankj.utilcode.util.TimeUtils;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.List;

/**
 * @author Aaron aaronzzxup@gmail.com
 */
class BookmarkAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int TYPE_EMPTY = 1;

    private List<Bookmark> mBookmarks;

    private DateFormat mDateFormat;

    @SuppressLint("SimpleDateFormat")
    BookmarkAdapter(List<Bookmark> list) {
        mBookmarks = list;
        mDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm");
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        LayoutInflater inflater = LayoutInflater.from(context);
        if (viewType == TYPE_EMPTY) {
            View itemView = inflater.inflate(R.layout.app_recycler_item_emptyview, parent, false);
            return new EmptyHolder(itemView);
        }
        View itemView = inflater.inflate(R.layout.app_recycler_item_bookmark, parent, false);
        ViewHolder holder = new ViewHolder(itemView);
        holder.itemView.setOnClickListener(v -> {
            int pos = holder.getAdapterPosition();
            Bookmark bk = mBookmarks.get(pos);
            ((IActivityComm) context).onJumpTo(bk.getPageId());
        });
        return holder;
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder viewHolder, int position) {
        if (!mBookmarks.isEmpty() && viewHolder instanceof ViewHolder) {
            ViewHolder holder = (ViewHolder) viewHolder;
            Bookmark bk = mBookmarks.get(position);
            holder.tvTitle.setText(bk.getTitle());
            holder.tvPageId.setText(String.valueOf(bk.getPageId() + 1));
            String time = TimeUtils.millis2String(bk.getTime(), mDateFormat);
            holder.tvTime.setText(time);
        } else if (viewHolder instanceof EmptyHolder) {
            EmptyHolder holder = (EmptyHolder) viewHolder;
            holder.itvEmpty.setVisibility(View.VISIBLE);
            holder.itvEmpty.setText(R.string.app_have_no_add_bookmark);
            holder.itvEmpty.setIconTop(R.drawable.app_ic_bookmark_emptyview);
        }
    }

    @Override
    public int getItemCount() {
        if (mBookmarks.isEmpty()) {
            return 1;
        }
        return mBookmarks.size();
    }

    @Override
    public int getItemViewType(int position) {
        if (mBookmarks.isEmpty()) {
            return TYPE_EMPTY;
        }
        return super.getItemViewType(position);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        TextView tvTitle, tvPageId, tvTime;

        ViewHolder(@NonNull View itemView) {
            super(itemView);
            tvTitle = itemView.findViewById(R.id.app_tv_title);
            tvPageId = itemView.findViewById(R.id.app_tv_page_id);
            tvTime = itemView.findViewById(R.id.app_tv_time);
        }
    }

    static class EmptyHolder extends RecyclerView.ViewHolder {
        ImageTextView itvEmpty;

        EmptyHolder(@NonNull View itemView) {
            super(itemView);
            itvEmpty = itemView.findViewById(R.id.app_itv_placeholder);
        }
    }
}

File: app/src/main/java/com/aaron/yespdf/preview/BookmarkAdapter.kt
package com.aaron.yespdf.preview

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.aaron.yespdf.R
import com.aaron.yespdf.common.EmptyHolder
import com.blankj.utilcode.util.TimeUtils
import kotlinx.android.synthetic.main.app_recycler_item_bookmark.view.*
import java.text.DateFormat
import java.text.SimpleDateFormat

/**
 * @author Aaron aaronzzxup@gmail.com
 */
internal class BookmarkAdapter(private val mBookmarks: List<Bookmark>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val mDateFormat: DateFormat

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val context = parent.context
        val inflater = LayoutInflater.from(context)
        if (viewType == TYPE_EMPTY) {
            val itemView = inflater.inflate(R.layout.app_recycler_item_emptyview, parent, false)
            return EmptyHolder(itemView)
        }
        val itemView = inflater.inflate(R.layout.app_recycler_item_bookmark, parent, false)
        val holder = ViewHolder(itemView)
        holder.itemView.setOnClickListener {
            val pos = holder.adapterPosition
            val bk = mBookmarks[pos]
            (context as IActivityInterface).onJumpTo(bk.pageId)
        }
        return holder
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        if (mBookmarks.isNotEmpty() && viewHolder is ViewHolder) {
            val bk = mBookmarks[position]
            viewHolder.itemView.app_tv_title.text = bk.title
            viewHolder.itemView.app_tv_page_id.text = (bk.pageId + 1).toString()
            val time = TimeUtils.millis2String(bk.time, mDateFormat)
            viewHolder.itemView.app_tv_time.text = time
        } else if (viewHolder is EmptyHolder) {
            viewHolder.itvEmpty.visibility = View.VISIBLE
            viewHolder.itvEmpty.setText(R.string.app_have_no_add_bookmark)
            viewHolder.itvEmpty.setIconTop(R.drawable.app_ic_bookmark_emptyview)
        }
    }

    override fun getItemCount(): Int {
        return if (mBookmarks.isEmpty()) {
            1
        } else mBookmarks.size
    }

    override fun getItemViewType(position: Int): Int {
        return if (mBookmarks.isEmpty()) {
            TYPE_EMPTY
        } else super.getItemViewType(position)
    }

    internal class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

    companion object {
        private const val TYPE_EMPTY = 1
    }

    init {
        mDateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm")
    }
}